File: java/parser/src/main/java/gov/lbl/parser/parser/RelationVisitor.java
package gov.lbl.parser.parser;

import java.util.List;

import gov.lbl.antlr4.visitor.modelicaBaseVisitor;
import gov.lbl.antlr4.visitor.modelicaParser;
import gov.lbl.parser.domain.Relation;
import gov.lbl.parser.domain.Arithmetic_expression;

import static java.util.stream.Collectors.toList;

public class RelationVisitor extends modelicaBaseVisitor<Relation> {
    @Override
    public Relation visitRelation(modelicaParser.RelationContext ctx) {
        Arithmetic_expressionVisitor arithmetic_expressionVisitor = new Arithmetic_expressionVisitor();
        List<Arithmetic_expression> arithmetic_expressions = ctx.arithmetic_expression()
                .stream()
                .map(arithmetic_expression -> arithmetic_expression.accept(arithmetic_expressionVisitor))
                .collect(toList()); 
        Rel_opVisitor rel_opVisitor = new Rel_opVisitor();
        String rel_op = ctx.rel_op() == null ? "" : ctx.rel_op().accept(rel_opVisitor);  
        
        Arithmetic_expression arithmetic_expression1 = arithmetic_expressions.get(0);
        Arithmetic_expression arithmetic_expression2 = null;
        if (arithmetic_expressions.size() > 1) {
            arithmetic_expression2 = arithmetic_expressions.get(1);
        }
        
        return new Relation(arithmetic_expression1, rel_op, arithmetic_expression2);
    }
}

File: jsParser/parser/RelationVisitor.js
const modelicaVisitor = require('../antlrFiles/modelicaVisitor').modelicaVisitor
const Relation = require('../domain/Relation')

const Arithmetic_expressionVisitor = require('./Arithmetic_expressionVisitor')
const Rel_opVisitor = require('./Rel_opVisitor')

class RelationVisitor {
  constructor () {
    modelicaVisitor.call(this)
    return this
  }

  visitRelation (ctx) {
    let arithmetic_expression1 = null
    let arithmetic_expression2 = null
    const arithmetic_expressions = []
    let rel_op = ''

    if (ctx.arithmetic_expression()) {
      const arithmetic_expressionVisitor = new Arithmetic_expressionVisitor.Arithmetic_expressionVisitor()
      ctx.arithmetic_expression().forEach(expr => {
        arithmetic_expressions.push(arithmetic_expressionVisitor.visitArithmetic_expression(expr))
      })
    }

    if (arithmetic_expressions.length > 0) {
      arithmetic_expression1 = arithmetic_expressions[0]
      if (arithmetic_expressions.length > 1) {
        arithmetic_expression2 = arithmetic_expressions[1]
      }
    }

    if (ctx.rel_op()) {
      const rel_opVisitor = new Rel_opVisitor.Rel_opVisitor()
      rel_op = rel_opVisitor.visitRel_op(ctx.rel_op())
    }

    return new Relation.Relation(arithmetic_expression1, rel_op, arithmetic_expression2)
  }
};

RelationVisitor.prototype = Object.create(modelicaVisitor.prototype)

exports.visitRelation = this.visitRelation
exports.RelationVisitor = RelationVisitor
